# Тема: Исследование надёжности заёмщиков


Цель: Выяснить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. 

Заказчик — кредитный отдел банка. 
Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd

statistics = pd.read_csv('/datasets/data.csv')
#Вызовем первые 15 строк таблицы
(statistics.head(15))

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

In [2]:
#Вызовем нижние 10 строк таблицы
(statistics.tail(10))

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

In [3]:
#Вызовем 10 строк таблицы случайным образом
(statistics.sample(10))

children  days_employed  dob_years education  education_id  \
1229          2            NaN         32   среднее             1   
5784          1   -1977.408685         30   среднее             1   
3641          1    -370.599364         45    высшее             0   
5451          2   -2946.602893         32    высшее             0   
11513         0    -755.499459         36   Среднее             1   
14495         0  380240.533813         53   СРЕДНЕЕ             1   
15693         0    -795.339974         25   СРЕДНЕЕ             1   
6744          1   -1613.781817         52    высшее             0   
1405          2   -2498.689711         33   среднее             1   
1126          1   -5078.763451         42   среднее             1   

               family_status  family_status_id gender income_type  debt  \
1229         женат / замужем                 0      F   сотрудник     0   
5784         женат / замужем                 0      M   сотрудник     0   
3641   Не женат / не замужем                 4      F   сотрудник     0   
5451         женат / замужем                 0      M   компаньон     0   
11513        женат / замужем                 0      F   сотрудник     0   
14495       гражданский брак                 1      F   пенсионер     0   
15693  Не женат / не замужем                 4      M   сотрудник     0   
6744         женат / замужем                 0      M   компаньон     0   
1405         женат / замужем                 0      F   сотрудник     0   
1126         женат / замужем                 0      M   сотрудник     0   

        total_income                                 purpose  
1229             NaN  строительство собственной недвижимости  
5784    93043.722711           получение высшего образования  
3641   103105.480918                            ремонт жилью  
5451   148000.834094                    покупка своего жилья  
11513   80981.039244                              автомобиль  
14495  110085.623272                                 свадьба  
15693  153536.867711                    покупка своего жилья  
6744   244888.266848           получение высшего образования  
1405   110915.752719                   получение образования  
1126   203230.987082                    сделка с автомобилем

In [4]:
#Вызовем общую информацию о таблице
statistics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
#Проверим корректность наименования столбцов
statistics.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

    В таблице 21525 строк и 12 столбцов. Типы данных у столбцов разные: int64, float64, object.
Столбцы содержат следующую информацию:
•	children — количество детей в семье
•	days_employed — общий трудовой стаж в днях
•	dob_years — возраст клиента в годах
•	education — уровень образования клиента
•	education_id — идентификатор уровня образования
•	family_status — семейное положение
•	family_status_id — идентификатор семейного положения
•	gender — пол клиента
•	income_type — тип занятости
•	debt — имел ли задолженность по возврату кредитов
•	total_income — ежемесячный доход
•	purpose — цель получения кредита
    
    Каждая строка таблицы содержит информацию о клиенте, с указанием его пола, возраста, уровня образования, типа занятости, трудового стажа, ежемесячного дохода, семейного положения, количества детей, цели получения кредита и  фактов наличия задолженности по возврату кредитов.  Для проверки рабочих гипотез необходимо использовать данные в первую очередь из столбцов family_status, children и debt.В первом содержатся категориальные данные, во втором - количественные, в третьем - логические.  В столбце total_income указан ежемесячный доход клиентов, в	purpose — цель получения кредита. 
     
**Вывод**: 
    
    Различное количество значений в столбцах указывает на то, что в данных есть пропущенные значения.  Названия столбцов корректны. В столбце days_employed имеются отрицательные значения, что не соотносится с трудовым стажем в днях, также видны пропущенные данные в этом столбце. Данные нуждаются в предобработке.
 
**Проблемы в рамках предобработки данных, которые нужно решать, исходя из полученной на текущем этапе информации:** 
 - работа с пропусками
 - проверка на наличие дубликатов
 - замена отрицательных значений в столбце со стажем
 - приведение к нижнему регистру данных в столбце с уровнем образования

## Шаг 2. Предобработка данных

### Обработка пропусков

In [6]:
#Проверим наличие пропусков во всей таблице.
(statistics.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [7]:
# Чтобы заменить пропуски на маркеры с отрицательным значением,
# предварительно нужно избавиться от изначально имеющихся отрицательных чисел в столбце с трудовым стажем. 
# Заменим при помощи abs() данные на абсолютную величину
statistics['days_employed'] = statistics['days_employed'].abs()

In [8]:
# Проверим, что получилось
(statistics['days_employed'].unique())

array([8437.67302776, 4024.80375385, 5623.42261023, ..., 2113.3468877 ,
       3112.4817052 , 1984.50758853])

In [9]:
#Заменим пропуски в столбцах. Так как значения колличественные, 
#используем fillna(), заменим неизвестный трудовой стаж на - 1000, так как менять 
#на среднее арифметическое, ноль или медиану в этом столбце без дополнительных действий некорректно
statistics['days_employed'] = statistics['days_employed'].fillna(value=- 1000)

In [10]:
# Проверим, есть ли пропуски
(statistics['days_employed'].isna().sum())

0

In [11]:
# Проверим, есть ли маркеры
(statistics['days_employed'].head(15))

0       8437.673028
1       4024.803754
2       5623.422610
3       4124.747207
4     340266.072047
5        926.185831
6       2879.202052
7        152.779569
8       6929.865299
9       2188.756445
10      4171.483647
11       792.701887
12     -1000.000000
13      1846.641941
14      1844.956182
Name: days_employed, dtype: float64

**Вывод по столбцу "Трудовой стаж в днях":**
    
    Мы удалили в трудовом стаже отрицательные значения, непонятно, откуда они взялись, возможно, некорректный ввод данных, или сбой при передаче данных.
    Считать трудовой стаж в днях неудобно, пропуски в нем заменены на маркеры со знаком минус, но нам это сейчас не важно, оставим как есть, эти данные мы не будем использовать.

In [12]:
#Вызовем к столбцам метод value_counts(), который возвращает уникальные значения и количество их упоминаний,
#определим уникальные значения в столбце с типом занятости.
(statistics['income_type'].value_counts())

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

Данные в столбце с типами занятости корректны. Можно их использовать без обработки.

In [13]:
# В столбце с уровнем дохода пропуски заменим на медиану в соответствующей категории типа занятости клиента
statistics['total_income'] = statistics.groupby('income_type')['total_income'].transform(lambda x: x.fillna(x.median()))
# Проверим, есть ли пропуски
(statistics['total_income'].isna().sum())

0

**Вывод:** 
    
    Пропуски имеются в столбцах с указанием трудового стажа и ежемесячного дохода в равном количестве, вероятно неслучайно, скорее всего пропущены данные при выгрузке. 
    Заменили значения NaN в столбцe days_employed на *маркер - 1000*, так как в нем колличественные данные, а заменять на ноль, среднее арифметическое, медиану или моду некорректно в данном случае, нужна дополнительная обработка данных.
    Заменили значения NaN в столбцe total_income на медиану в соответствующей категории по типу занятости клиента, так как в нем колличественные данные, данные мы будем использовать, заменять на среднее арифметическое, ноль или моду некорректно в данном случае. 

**Проверим столбцы на корректность имеющихся данных** 

In [14]:
#Вызовем к столбцам метод value_counts(), который возвращает уникальные значения и количество их упоминаний,
#определим уникальные значения в столбце с количеством детей.
(statistics['children'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

**Вывод по столбцу "Количество детей":**
    
    Имеются отрицательные значения (47), количество детей не может быть отрицательным, необходимо заменить данные.
    Имеются данные, что у 76 клиентов по 20 детей. Крайне маловероятно. Скорее всего, ошибка, необходимо заменить данные. 

In [15]:
# Заменим -1 на 0, выбор основан на "моде", она в этом столбце равна 0.
statistics['children'] = statistics['children'].replace(-1, 0)

In [16]:
# Заменим 20 на 2, возможно случайно был поставлен ноль рядом с 2, учитывая выбор замены -1 на 0, 
#статистика не должна пострадать.
statistics['children'] = statistics['children'].replace(20,2)

In [17]:
#Проверим результат
(statistics['children'].value_counts())

0    14196
1     4818
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Столбец 'children' в порядке.

In [18]:
#Проверим сколько клиентов определенного возраста берут кредит, сразу сгруппируем по увеличению старшинства
(statistics.groupby('dob_years')['dob_years'].count())

dob_years
0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

**Вывод по столбцу "Возраст клиента":**

    101 клиент недавно родился. Данные были либо потеряны и заменены на ноль, либо ошибки при внесении данных. В данном случае ничего сделать нельзя, оставляем как есть, на результат исследования это не повлияет. 

In [19]:
#Вызовем к столбцам метод value_counts(), который возвращает уникальные значения и количество их упоминаний,
#определим уникальные значения в столбце с уровнем образования.
(statistics['education'].value_counts())

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

**Вывод по столбцу "Уровень образования":**

Очевидно, что из-за разного написания у нас гораздо больше уровней образования, чем в принципе возможно в данном случае. Необходимо данные привести к нижнему регистру.

In [20]:
# Приведем строки к нижнему регистру
statistics['education'] = statistics['education'].str.lower()
#Проверим результат
(statistics['education'].value_counts())

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Столбец 'education' в порядке.

In [21]:
#Вызовем к столбцам метод value_counts(), который возвращает уникальные значения и количество их упоминаний,
#определим уникальные значения в столбце с семейным положением.
(statistics['family_status'].value_counts())

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

**Вывод по столбцу "Семейное положение":**

    Ничего критичного не выявлено, но единообразие тут уместно, данные приведем к нижнему регистру.

In [22]:
# Приведем строки к нижнему регистру
statistics['family_status'] = statistics['family_status'].str.lower()
#Проверим результат
(statistics['family_status'].value_counts())

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Столбец 'family_status' в порядке.

In [23]:
#Вызовем к столбцам метод value_counts(), который возвращает уникальные значения и количество их упоминаний,
#определим уникальные значения в столбце с полом.
(statistics['gender'].value_counts())

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

**Вывод по столбцу "Пол":**
    
    Какой - то неизвестный науке пол. Можно выяснить, где он  и заменить на "моду".

In [24]:
# Заменим XNA на "моду", ведь нет никаких данных об имени клиента, чтобы узнать его настоящий пол
statistics['gender'] = statistics['gender'].replace('XNA', 'F')
# Проверим
(statistics['gender'].value_counts())

F    14237
M     7288
Name: gender, dtype: int64

Столбец 'gender' в целом в порядке, условно среди клиентов на одну женщину больше, чем точно известно. Можно было оставить все без изменений, но судя по данным, женщины почти в 2 раза чаще берут кредит.

In [25]:
#Вызовем к столбцам метод value_counts(), который возвращает уникальные значения и количество их упоминаний,
#определим уникальные значения в столбце с данными о задолженностях.
(statistics['debt'].value_counts())

0    19784
1     1741
Name: debt, dtype: int64

**Вывод по столбцу "Факт просрочки платежа, наличие задолженностей":**

   Данные корректны.

In [26]:
# Разберемся окончательно с ежемесячным доходом,  
# проверим уникальные значения unique()()
(statistics['total_income'].unique())

array([253875.6394526 , 112080.01410244, 145885.95229686, ...,
        89672.56115303, 244093.05050043,  82047.41889948])

In [27]:
# Строк слишком много, 
# проверим количество уникальных значений nunique()()
(statistics['total_income'].nunique())

19353

**Вывод по столбцу "Ежемесячный доход":**

Не ясна величина измерения дохода. Не ясно, это доход заемщика, или в целом на семью. Эти сведения необходимо запрашивать у заказчиков.
Имеются странные точки. Пусть остаются.  

In [28]:
#Вызовем к столбцам метод value_counts(), который возвращает уникальные значения и количество их упоминаний,
#определим уникальные значения в столбце с данными о целях кредита.
(statistics['purpose'].value_counts())

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

**Вывод по столбцу "Цель кредита":**

Большое разнообразие желаний вынуждает людей обращаться за кредитом в банки. Однако некоторые цели можно привести к единой формулировке, настолько они близки по смыслу. Займемся этим в разделе лемматизации.

### Замена типа данных

In [29]:
#Заменим вещественный тип данных на целочисленный, удобно использовать astype('int').
statistics['days_employed'] = statistics['days_employed'].astype('int')
statistics['total_income'] = statistics['total_income'].astype('int') 
#Проверим результат вызовом общей информации о таблице.
statistics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**: 
    
    В столбцах total_income и days_employed теперь целые числа, с ними удобнее работать.

### Обработка дубликатов

In [30]:
#Проверим наличие и количество дубликатов в целом в таблице.
#Метод выбран исходя из данных, содержащихся в таблице
(statistics.duplicated().sum())

71

In [31]:
#Удалим 54 выявленных дубля, заменим нумерацию,проверим результат
statistics = statistics.drop_duplicates().reset_index(drop = True)
(statistics.duplicated().sum())


0

**Вывод**: 
    
    Выявили и удалили 54 дубля в таблице с данными. Вероятная причина появления дубликатов - неправильное соединение данных из разных источников, ошибки пользователя при занесении информации.

### Лемматизация

In [32]:
#Создадим новую переменную, в которой сохраним данные столбца с целями кредита
credit_purposes = statistics['purpose'].unique().tolist()
#Проверим результат
credit_purposes

['покупка жилья',
 'приобретение автомобиля',
 'дополнительное образование',
 'сыграть свадьбу',
 'операции с жильем',
 'образование',
 'на проведение свадьбы',
 'покупка жилья для семьи',
 'покупка недвижимости',
 'покупка коммерческой недвижимости',
 'покупка жилой недвижимости',
 'строительство собственной недвижимости',
 'недвижимость',
 'строительство недвижимости',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'операции с коммерческой недвижимостью',
 'строительство жилой недвижимости',
 'жилье',
 'операции со своей недвижимостью',
 'автомобили',
 'заняться образованием',
 'сделка с подержанным автомобилем',
 'получение образования',
 'автомобиль',
 'свадьба',
 'получение дополнительного образования',
 'покупка своего жилья',
 'операции с недвижимостью',
 'получение высшего образования',
 'свой автомобиль',
 'сделка с автомобилем',
 'профильное образование',
 'высшее образование',
 'покупка жилья для сдачи',
 'на покупку автомобиля',
 'ремонт жилью',
 '

In [33]:
#Переведем данные в строку
from pymystem3 import Mystem
m = Mystem()
string = '; '.join(credit_purposes)
(string)

'покупка жилья; приобретение автомобиля; дополнительное образование; сыграть свадьбу; операции с жильем; образование; на проведение свадьбы; покупка жилья для семьи; покупка недвижимости; покупка коммерческой недвижимости; покупка жилой недвижимости; строительство собственной недвижимости; недвижимость; строительство недвижимости; на покупку подержанного автомобиля; на покупку своего автомобиля; операции с коммерческой недвижимостью; строительство жилой недвижимости; жилье; операции со своей недвижимостью; автомобили; заняться образованием; сделка с подержанным автомобилем; получение образования; автомобиль; свадьба; получение дополнительного образования; покупка своего жилья; операции с недвижимостью; получение высшего образования; свой автомобиль; сделка с автомобилем; профильное образование; высшее образование; покупка жилья для сдачи; на покупку автомобиля; ремонт жилью; заняться высшим образованием'

In [34]:
#Проведем лемматизацию значений
lemmas = m.lemmatize(string)
from collections import Counter
(Counter(lemmas))

Counter({'покупка': 10,
         ' ': 59,
         'жилье': 7,
         '; ': 37,
         'приобретение': 1,
         'автомобиль': 9,
         'дополнительный': 2,
         'образование': 9,
         'сыграть': 1,
         'свадьба': 3,
         'операция': 4,
         'с': 5,
         'на': 4,
         'проведение': 1,
         'для': 2,
         'семья': 1,
         'недвижимость': 10,
         'коммерческий': 2,
         'жилой': 2,
         'строительство': 3,
         'собственный': 1,
         'подержать': 2,
         'свой': 4,
         'со': 1,
         'заниматься': 2,
         'сделка': 2,
         'получение': 3,
         'высокий': 3,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1,
         '\n': 1})

*Используя выделенные леммы существительных можно определить следующие основные цели кредита:*
1) Недвижимость/ жилье
2) Автомобиль
3) Образование
4) Свадьба

In [35]:
# Создадим список новых категорий по цели кредита
categories = ["недвижимость", "жилье", "автомобиль", "образование", "свадьба"]

# Проведем лемматизацию,
# одновременно заменим список лемм в каждой строке на главное ключевое слово из списка категорий

def lemmatize(text):
    lemma = m.lemmatize(text)
    for word in categories:
        if word in lemma:
            lemma = word
    return lemma
# Создадим новый столбец с категориями по целям кредита
statistics['purpose_group'] = statistics['purpose'].apply(lemmatize)        
statistics['purpose_group'].value_counts().head(15)

недвижимость    6351
жилье           4460
автомобиль      4306
образование     4013
свадьба         2324
Name: purpose_group, dtype: int64

In [36]:
#Так как жилье относится к недвижимости, сделаем соответствующую замену
statistics.loc[statistics['purpose_group'] == 'жилье', 'purpose_group'] = 'недвижимость'
#Проверим
statistics['purpose_group'].value_counts()

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2324
Name: purpose_group, dtype: int64

**Вывод**
    
    Используя выделенные леммы существительных определили следующие основные цели кредита:
    1) недвижимость (включая жилье)
    2) автомобиль
    3) образование
    4) свадьба
    Создали соответствующий столбец в таблице. Похоже, мы уже приступили к категоризации данных.

### Категоризация данных

    Мы уже категоризировали цели кредита. Исходя из цели исследования, проведем категоризацию следующих данных:
- количество детей (для решения задачи исследования можно выделить всего две категории: бездетные, с детьми (ведь нас интересует влияет ли *наличие* детей на своевременность возврата кредита);
- уровень дохода (воспользуемся стандартным подходом, используем процентели в имеющихся данных: низкий 0-30%, средний 30-60%, выше среднего 60-90 %, высокий > 90 %; выборка порогов субъективна и однозначно условна);
- наличие задолженности по возврату кредитов (удобно видеть наглядно, у кого имеются просрочки платежей, очевидны две категории: должники, своевременный возврат кредита).
    
    Семейное положение категоризировать не будем, основные категории уже приведены в данных.

In [37]:
#Используем функцию для создания категорий по количеству детей
def children_status(children):
    if children > 0: return 'С детьми'
    else: return 'Бездетные'

#Применим метод, добавим новый столбец и проверим результат
statistics['children_status'] = statistics['children'].apply(children_status)
statistics['children_status'].value_counts()

Бездетные    14138
С детьми      7316
Name: children_status, dtype: int64

Данные по детям категоризированы.

In [38]:
#Определим процентили столбца с уровнем дохода, воспользуемся quantile()
(statistics['total_income'].quantile([0.30,0.60,0.90]))

0.3    116008.6
0.6    161335.0
0.9    269825.5
Name: total_income, dtype: float64

In [39]:
#Используем функцию для создания категорий по уровню дохода
def income_status(total_income):
    if total_income <= 116062.0:
            return 'Низкий уровень дохода'
    if total_income <= 156400.0:
            return 'Средний уровень дохода'
    if total_income <= 269784.0:
            return 'Выше среднего уровень дохода'
    return 'Высокий уровень дохода'
#Применим метод, добавим новый столбец и проверим результат
statistics['income_status'] = statistics['total_income'].apply(income_status)
statistics['income_status'].value_counts()

Выше среднего уровень дохода    6942
Низкий уровень дохода           6442
Средний уровень дохода          5922
Высокий уровень дохода          2148
Name: income_status, dtype: int64

Данные по доходу категоризированы.

In [40]:
#Используем функцию для создания категорий по своевременности возврата кредита
def debt_status(debt):
    if debt > 0: return 'Должники'
    else: return 'Своевременный возврат кредита'

#Применим метод, добавим новый столбец и проверим результат
statistics['debt_status'] = statistics['debt'].apply(debt_status)
statistics['debt_status'].value_counts()

Своевременный возврат кредита    19713
Должники                          1741
Name: debt_status, dtype: int64

Данные по своевременности возврата кредита категоризированы

**Вывод**
    Пора приступать непосредственно к исследованию. Данные подготовлены.

## Шаг 3. Ответьте на вопросы

- **Есть ли зависимость между наличием детей и возвратом кредита в срок?**

In [41]:
#Для наглядности удобно создать сводную таблицу, используем метод pivot_table()
children_pivot = statistics.pivot_table(index = ['children_status'], columns = ['debt'], values = 'debt_status', aggfunc='count')
#Создадим новый столбец, в котором отобразим процент должников по каждой категории,
#для большинства людей результат в процентах легче оценить, чем сравнивать отношение
children_pivot['debt_%'] = children_pivot[1] / (children_pivot[0] + children_pivot[1]) * 100
#Так как нас интересую в первую очередь неблагонадежные клиенты, отсортируем по убыванию
children_pivot.sort_values(by = 'debt_%', ascending = False)

debt                 0     1    debt_%
children_status                       
С детьми          6639   677  9.253691
Бездетные        13074  1064  7.525817

**Вывод**: В целом задолженность по выплате кредитов среди клиентов с детьми и бездетных находится примерно на одном уровне. Однако почти на 2% больше клиентов с детьми имеют задолженность по выплате кредита, что можно объяснить расстановкой приоритетов в социуме - в случае непредвиденных обстоятельств обеспечить насущные потребности детей важнее, чем вовремя погасить кредит.А также статистически не верной выборкой - в исходных данных в два раза больше бездетных клиентов.
В тоже время, наличие или отсутствие детей не может быть основным критерием благонадежности клиента. Количество детей по сути необходимо знать только для расчета приблизительного уровня расходов клиента.

- **Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [42]:
#Для наглядности удобно создать сводную таблицу, используем метод pivot_table()
family_status_pivot = statistics.pivot_table(index = ['family_status'], columns = ['debt'], values = 'debt_status', aggfunc='count')
#Создадим новый столбец, в котором отобразим процент должников по каждой категории,
#для большинства людей результат в процентах легче оценить, чем сравнивать отношение
family_status_pivot['debt_%'] = family_status_pivot[1] / (family_status_pivot[0] + family_status_pivot[1]) * 100
#Так как нас интересую в первую очередь неблагонадежные клиенты, отсортируем по убыванию
family_status_pivot.sort_values(by = 'debt_%', ascending = False)

debt                       0    1    debt_%
family_status                              
не женат / не замужем   2536  274  9.750890
гражданский брак        3763  388  9.347145
женат / замужем        11408  931  7.545182
в разводе               1110   85  7.112971
вдовец / вдова           896   63  6.569343

**Вывод**: Неплательщики имеются среди людей всех семейных статусов. Однако среди юридически свободных от уз брака людей (не женат/не замужем, гражданский брак) чаще встречаются просрочки платежей (разница примерно в 2 % клиентов). Опыт семейной жизни (женат/замужем, в разводе, вдовец/вдова) вырабатывает в людях ответственность и чувство долга, а также умение правильно планировать свой бюджет, поэтому такие клиенты реже попадают в категорию должников. 
Несмотря на незначительную разницу между количеством должников в разных категориях, все же учитывать семейный статус человека при  оценке благонадежности клиента следует, ведь социальные риски сильнее проявляются на одиноких/свободных людях. Отсутствие обязательств в такой социально значимой сфере как брак откладывает отпечаток на отношение к обязательствам в других сферах жизни.

- **Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [43]:
#Для наглядности удобно создать сводную таблицу, используем метод pivot_table()
income_status_pivot = statistics.pivot_table(index = ['income_status'], columns = ['debt'], values = 'debt_status', aggfunc='count')
#Создадим новый столбец, в котором отобразим процент должников по каждой категории,
#для большинства людей результат в процентах легче оценить, чем сравнивать отношение
income_status_pivot['debt_%'] = income_status_pivot[1] / (income_status_pivot[0] + income_status_pivot[1]) * 100
#Так как нас интересую в первую очередь неблагонадежные клиенты, отсортируем по убыванию
income_status_pivot.sort_values(by = 'debt_%', ascending = False)

debt                             0    1    debt_%
income_status                                    
Средний уровень дохода        5406  516  8.713273
Низкий уровень дохода         5918  524  8.134120
Выше среднего уровень дохода  6392  550  7.922789
Высокий уровень дохода        1997  151  7.029795

**Вывод**: Люди с высоким уровнем дохода реже попадают в категорию должников, на 1 % реже, чем с доходом выше среднего. Это можно объяснить тем, что данная категория клиентов наиболее социально стабильна, с одной стороны они обеспечены и защищены от наиболее распростаненных рисков, с другой - не играют "по - крупному", не входят в категорию высокообеспеченных людей, в которой действуют свои законы и свои риски. 
Клиенты со средним уровнем дохода чаще являются должниками, однако разница с должниками среди клиентов с относительно низким уровнем дохода незначительна, вероятно это связано с относительным характером выборки  для данных категорий.
Уровень доходов является основным критерием для оценки платежеспособности клиента. Однако благонадежность клиента зависит в первую очередь не от дохода, а от базовых качеств его личности и жизненных обстоятельств.
Нельзя аргументировано утверждать, что уровень дохода влияет на его надежность как клиента банка, уровень дохода влияет только на вероятную возможность оплатить кредит вовремя.

- **Как разные цели кредита влияют на его возврат в срок?**

In [44]:
#Для наглядности удобно создать сводную таблицу, используем метод pivot_table()
purpose_group_pivot = statistics.pivot_table(index = ['purpose_group'], columns = ['debt'], values = 'debt_status', aggfunc='count')
#Создадим новый столбец, в котором отобразим процент должников по каждой категории,
#для большинства людей результат в процентах легче оценить, чем сравнивать отношение
purpose_group_pivot['debt_%'] = purpose_group_pivot[1] / (purpose_group_pivot[0] + purpose_group_pivot[1]) * 100
#Так как нас интересую в первую очередь неблагонадежные клиенты, отсортируем по убыванию
purpose_group_pivot.sort_values(by = 'debt_%', ascending = False)

debt               0    1    debt_%
purpose_group                      
автомобиль      3903  403  9.359034
образование     3643  370  9.220035
свадьба         2138  186  8.003442
недвижимость   10029  782  7.233373

**Вывод**: Наиболее благонадежны в плане своевременности возврата долга клиенты, берущие кредиты на приобретение, постройку жилья и недвижимости, сюда же входят люди, инвестирующие в недвижимость. Очевидно, это связано со степенью серьезности намерений клиента - важная цель требует ответственного подхода.
Наибольшее количество должников среди клиентов с целями приобретения, ремонта атомобиля. В данную группу безусловно входят клиенты, берущие кредит без основательной оценки своих возможностей по его погашению - желание передвигаться комфортно и повысить свой статус в ближайшем окружении, особенно среди молодых людей, не способствует холодному расчету. Данная цель однозначно в зоне риска по благонадежности клиента.
Примерно на этом же уровне находится взятие кредита в целях получения, повышения уровня образования. Здесь риски связаны с тем, что как правило данный кредит необходим молодым людям или для молодых людей. Сложно оценить благонадежность клиента в данном случае, риски изменения жизненных обстоятельст велики. 
Кредит на свадьбу своевременно оплачивают чаще, чем кредиты на автомобиль и образование, вероятно это связано с относительно небольшими суммами кредита.

## Шаг 4. Общий вывод

   **Исходя из имеющихся исходных данных можно выделить следующие основные результаты исследования:**
   - Наличие детей незначительно влияет на благонадежность клиентов, бездетные клиенты чаще возвращают кредит своевременно в сравнении с клиентами, имеющими детей. Разница составляет около 2 % клиентов. Следует учитывать, что в представленных для исследования данных выборка бездетных в два раза выше, чем клиентов с детьми. Нельзя считать полученные выводы статистически верными.
   - Семейное положение влияет на своевременность возврата кредитов. Люди с опытом семейной жизни более благонадежны.
   - Уровень дохода не влияет на своевременность возврата кредита.
   - Цель кредита заметно влияет на своевременность его возврата. Наиболее выгодная для банка в этом плане цель кредита - операции с жильем, недвижимостью.

   **Рекомендации для заказчика**
   Рассмотреть вопрос о функциональной эффективности системы для сбора данных о клиентах. Минимизировать ручной ввод информации пользователями системы сбора данных в целях повышения уровня корректности вносимых сведений (выбор возраста, пола и т.п.), блокировать сохранение при отсутствии ввода обязательных данных). Максимально категоризировать собираемые данные. Проанализировать существующую систему мотивации сотрудников, ответственных за корректный сбор и внесение данных.
   Собрать данные для приблизительной оценки расходов клиента (использовать потребительскую корзину и реакльные расходы населения), с учетом доходов и расходов всех членов его семьи. Исследовать зависимость возврата кредита в срок от расходов клиента. 
   Исследовать зависимость благонадежности клиента от его пола, образования, типа занятости, возраста.
    